src/Data/Expr.purs

```diff
@@ -2,89 +2,389 @@ module Lynx.Data.Expr where
 
 import Prelude
 
+import Control.Alt ((<|>))
 import Data.Argonaut (class DecodeJson, class EncodeJson, Json, decodeJson, encodeJson, jsonEmptyObject, stringify, (.:), (:=), (~>))
 import Data.BigInt as Data.BigInt
 import Data.DateTime (DateTime)
 import Data.Either (Either(..))
-import Data.Foldable (fold, intercalate)
+import Data.Eq (class Eq1)
+import Data.Foldable (class Foldable, fold, foldlDefault, foldrDefault)
 import Data.Formatter.DateTime (format, unformat)
 import Data.Formatter.Parser.Interval (extendedDateTimeFormatInUTC)
+import Data.Functor.Coproduct.Inject (inj, prj)
+import Data.Functor.Coproduct.Nested (type (<\/>), (<\/>))
+import Data.Functor.Mu (Mu)
 import Data.Generic.Rep (class Generic)
 import Data.Generic.Rep.Show (genericShow)
-import Data.Maybe (Maybe(..), maybe')
-import Data.Newtype (wrap)
+import Data.Maybe (Maybe(..), maybe, maybe')
+import Data.Newtype (class Newtype, un, wrap)
 import Data.NonEmpty (NonEmpty(..))
 import Data.String.Common (replace)
 import Data.String.Pattern (Pattern(..), Replacement(..))
+import Data.Traversable (class Traversable, traverse)
 import Foreign.Object (Object)
 import Foreign.Object as Foreign.Object
+import Matryoshka (class Corecursive, class Recursive, anaM, cata, embed, project)
 import Ocelot.Data.Currency (Cents, formatCentsToStrDollars)
 import Test.QuickCheck (class Arbitrary, arbitrary)
 import Test.QuickCheck.Gen (Gen, Size, arrayOf, oneOf, sized)
 
 type Key = String
 
-data ExprType
-  = Array (Array ExprType)
-  | Boolean Boolean
-  | Cents Cents
-  | DateTime DateTime
-  | Int Int
-  | Pair { name :: ExprType, value :: ExprType }
-  | String String
+newtype ArrayF a
+  = Array (Array a)
+
+derive instance eqArrayF :: (Eq a) => Eq (ArrayF a)
+
+derive instance eq1ArrayF :: Eq1 ArrayF
+
+derive instance functorArrayF :: Functor ArrayF
+
+derive newtype instance showArrayF :: (Show a) => Show (ArrayF a)
+
+derive newtype instance foldableArrayF :: Foldable ArrayF
+
+derive newtype instance traversableArrayF :: Traversable ArrayF
+
+decodeArrayF :: ExprJSON -> Either String (ArrayF ExprJSON)
+decodeArrayF { out, param } = case out of
+  "Array" -> map Array (decodeJson param)
+  _ -> Left "Expected Array"
+
+encodeArrayF :: ArrayF ExprJSON -> ExprJSON
+encodeArrayF = case _ of
+  Array x -> base "Array" x
+
+isEmptyArrayF :: forall a. ArrayF a -> Boolean
+isEmptyArrayF = case _ of
+  Array [] -> true
+  Array _ -> false
+
+printArrayF :: ArrayF String -> String
+printArrayF = case _ of
+  Array x -> show x
```

Comment:
So the point of `print`, at least as I intended, was to cast our types to strings for the purpose of concating, ultimately for the purpose of printing to the screen. My original motivation, when still working with primitives, was something like `show`, but that wouldn't wrap quotes around strings. That way concating `"foo"` and `"bar"` wouldn't produce `""foo""bar""`, but just `"foobar"`.

Extending that idea to arrays, we probably just want a comma-separated string of array values. Something like
```purescript
Array x -> intercalate ", " (print <$> x)
```

Reply:
Are you okay with doing that change separate from this PR? Ideally we can add a test suite around `print` so we don't have to think about whether code changes are altering behavior. The changes in this PR don't alter the way `print` currently works. I.e. On both `master` and this branch `print (array_ [int_ 1, int_ 2, int_ 3]) == "[1, 2, 3]"`.

Reply:
Absolutely. I was actually thinking that this morning, that the `print` changes should be done separately. I like the idea of adding a test suite around `print`, since a type signature of `a -> String` doesn't have any guarantees about what the function of that string is